server/src/main/java/org/elasticsearch/cluster/routing/allocation/DiskThresholdSettings.java

```diff
@@ -130,143 +125,89 @@ public DiskThresholdSettings(Settings settings, ClusterSettings clusterSettings)
-        }
-
-    }
-
-    static final class HighDiskWatermarkValidator implements Setting.Validator<String> {
-
-        @Override
-        public void validate(final String value) {
-
-        }
 
-        @Override
-        public void validate(final String value, final Map<Setting<?>, Object> settings) {
-            final String lowWatermarkRaw = (String) settings.get(CLUSTER_ROUTING_ALLOCATION_LOW_DISK_WATERMARK_SETTING);
-            final String floodStageRaw = (String) settings.get(CLUSTER_ROUTING_ALLOCATION_DISK_FLOOD_STAGE_WATERMARK_SETTING);
-            doValidate(lowWatermarkRaw, value, floodStageRaw);
+            if (low.isAbsolute() == false && high.isAbsolute() == false && flood.isAbsolute() == false) { // Validate as percentages
+                final double lowWatermarkThreshold = low.getRatio().getAsPercent();
+                final double highWatermarkThreshold = high.getRatio().getAsPercent();
+                final double floodThreshold = flood.getRatio().getAsPercent();
+                if (lowWatermarkThreshold > highWatermarkThreshold) {
+                    throw new IllegalArgumentException(
+                        "low disk watermark [" + low.getStringRep() + "] more than high disk watermark [" + high.getStringRep() + "]"
+                    );
+                }
+                if (highWatermarkThreshold > floodThreshold) {
+                    throw new IllegalArgumentException(
+                        "high disk watermark ["
+                            + high.getStringRep()
+                            + "] more than flood stage disk watermark ["
+                            + flood.getStringRep()
+                            + "]"
+                    );
+                }
+            } else if (low.isAbsolute() && high.isAbsolute() && flood.isAbsolute()) { // Validate as absolute values
+                final ByteSizeValue lowWatermarkBytes = low.getAbsolute();
+                final ByteSizeValue highWatermarkBytes = high.getAbsolute();
+                final ByteSizeValue floodStageBytes = flood.getAbsolute();
+
+                if (lowWatermarkBytes.getBytes() < highWatermarkBytes.getBytes()) {
+                    throw new IllegalArgumentException(
+                        "low disk watermark [" + low.getStringRep() + "] less than high disk watermark [" + high.getStringRep() + "]"
+                    );
+                }
+                if (highWatermarkBytes.getBytes() < floodStageBytes.getBytes()) {
+                    throw new IllegalArgumentException(
+                        "high disk watermark ["
+                            + high.getStringRep()
+                            + "] less than flood stage disk watermark ["
+                            + flood.getStringRep()
+                            + "]"
+                    );
+                }
+            } else {
+                final String message = String.format(
+                    Locale.ROOT,
```

Comment:
nit (not introduced in this change but since we're touching this line anyway...):

```suggestion
                final String message = Strings.format(
```

Reply:
@DaveCTurner had to keep this because it produced this:

```
> Task :server:forbiddenApisMain
Forbidden method invocation: java.lang.String#format(java.lang.String,java.lang.Object[]) [Uses default locale]
  in org.elasticsearch.cluster.routing.allocation.DiskThresholdSettings$WatermarkValidator (DiskThresholdSettings.java:185)
Scanned 7046 class file(s) for forbidden API invocations (in 2.12s), 1 error(s).

> Task :server:forbiddenApisMain FAILED
```

Please unresolve conversation if this is not sufficient.

Reply:
Yes you mustn't use `java.lang.String#format` but `org.elasticsearch.core.Strings#format` is preferable (it hard-codes the `Locale` and is exception-safe too).

Reply:
Thanks! done. Please unresolve conversation is it is not sufficient.